webapp/sources/ldap-inventory/inventory-provisioning-web/src/main/scala/com/normation/inventory/provisioning/endpoint/InventoryFileWatcher.scala

```diff
@@ -211,44 +214,106 @@ class InventoryFileWatcher(
+   * We only modify that map as a result of a dequeue event.
    */
-  private var toBeProcessed = Map.empty[File, SerialCancelable]
-
-  private def remove(file: File): Unit = synchronized {
-    toBeProcessed = (toBeProcessed - file)
-  }
+  val toBeProcessed = ZioRuntime.unsafeRun(scalaz.zio.RefM.make(Map.empty[File, Fiber[RudderError, Unit]]))
 
-  val newTask = (f:File) => Task (processFile(f) ).delayExecution(500 millis).doOnFinish (_ => Task (remove(f)))
+  /*
+   * We need a queue of add file / file written even to delimitate when a file should be
+   * processed.
+   * We are going to receive a streem of "write file xxx" event, and updating the map accordingly
+   * (reseting the task). Once the task is started, we want to free space in the map, and so enqueue
+   * a "file written" event.
+   */
+  type WatchQueue = ZQueue[Any, Nothing, Any, Nothing, WatchEvent, WatchEvent]
+  val queue = ZioRuntime.unsafeRun(Queue.bounded[WatchEvent](10000))
+
+  // processing message is a loop, so UIO[Nothing]
+  def loop: UIO[Nothing] = processMessage().flatMap(_ => loop)
+  def processMessage(): UIO[Unit] = {
+    queue.take.flatMap {
+      case WatchEvent.End(file) => // simple case: remove file from map
+        toBeProcessed.update[Any](s =>
+          (s - file).succeed
+        ).unit
+
+      case WatchEvent.Mod(file) =>
+        // look if the file is already here. If so, interrupt. In all case, create a new task.
+        (toBeProcessed.update { s =>
+          val newMap = {
+            // the new task must :
+            // - wait for 500 ms for interruption
+            // - then execute on different IO scheduler
+            // - if the map wasn't interrupted in the first 500 ms, it is not interruptible anymore
+
+            val effect = ZioRuntime.Environment.blocking.blockingExecutor.flatMap(executor =>
+              (
+                   UIO.unit.delay(fileWrittenThreshold).provide(ZioRuntime.Environment)
+                *> queue.offer(WatchEvent.End(file))
+                *> processFile(file, locks).uninterruptible
+              ).fork.lock(executor)
+            )
+
+            effect.map(e => s + (file -> e))
+          }
 
-  def addFile(file: File): Unit =  synchronized {
-    toBeProcessed.get(file) match {
-      case None =>
-          // the delay here is to let some time pass to let the file be fully written
-          // the timeout is the max time allowed for file to be full written. Not that
-          // these files will be proccessed by agent run latter.
-          toBeProcessed = (toBeProcessed + (file -> SerialCancelable(newTask(file).runAsync)))
+          s.get(file) match {
+            case None =>
+              newMap
+            case Some(existing) =>
+              existing.interrupt.unit *> newMap
+          }
+        }).unit
+    }
+  }
 
-        case Some(ref) =>
-          ref := newTask(file).runAsync
+  //start the process
```

Comment:
The place where we start the process that continuously look for event in the loop.